### ISSUE TITLE
PyGobject on Windows needs modification of PATH envvar

### ISSUE THREAD

**Issue by andreubotella** [stuaxo/vext#31]
On Linux distros we're relying on the package manager to install the system-wide packages and the native libraries they depend on, so they're available in the OS's system path without dependency hell. Since Windows doesn't have a package manager, applications usually don't install dependencies system-wide, but on the application folder.

For native dependencies to Python packages it's the same, plus the question of where to place the dependencies. The installers for PyQt seem to do it practically but not very elegantly by dumping all the Qt libraries into the Python package directory – and this plays well with vext.

On the other hand, the PyGObject installer at Sourceforge has the Python package in `site-packages\gi` and the GTK+ libraries in `site-packages\gnome`, with a `rpath.pth` file that adds the latter path to `os.environ['PATH']`. And of course, vext is ignoring this file, with the result that the `gi` module in the virtualenv tries to load the GTK+ libraries from the PATH and finds nothing (`ImportError: DLL load failed: The specified module could not be found.`).

The contents of `rpath.pth`:

```
import distutils.sysconfig, sys, os; os.environ['PATH'] = os.path.join(distutils.sysconfig.get_python_lib(),'gnome;') + os.getenv('PATH')
```

The quick fix would be to install a custom `rpath.pth` along with vext.gi, making sure that it doesn't do anything if there's no `gnome` directory – GTK+ could still be installed in the system path. But this feels to prone to break if anything changes.

**Comment by stuaxo** [comment]
I tried adding it to the .pth section of gi.vext:

```
pths:
    rpath.pth
```

The path is added, but isn't correct yet:

```
C:\\test-venv\\Lib\\site-packages\\gnome
```

In our version of addpackage we'll need to check check if any directories added to the path are in the virtualenv site-packages and change them to point at the system site-packages.

**Comment by stuaxo** [comment]
Try it now ... it works for `import gi` - but I haven't had a chance to test further.

Thanks for the detailed bug report, it would have taken a lot longer to work out what was going on without it :)

**Comment by stuaxo** [comment]
Cairocffi doesn't seem to work - can't work out if it should or not with vext.gi + cairocffi ...

**Comment by szabolcsdombi** [comment]
Python 3.5 and 3.6 has an embedded version for windows where you could run **[get-pip](https://bootstrap.pypa.io/get-pip.py)**
placing *.pyd files next to the python.exe is working as well.

**Comment by stuaxo** [comment]
@cprogrammer1994 is this the right issue ?   Not quite sure I understand - this is about GI (not GL :) ) - which is the bindings for Gtk3 in python.

**Comment by szabolcsdombi** [comment]
i thought you are facing problems on windows because of the PATH variable. I mentioned an alternative to sandbox python.
**PEP 486**, making the Windows Python launcher aware of virtual environments (in 3.5.0)

**Comment by stuaxo** [comment]
Interesting, hadn't seen that pep.

Some Context:

The vext.\* projects are to allow you to use libraries in virtualenvs that are usually tricky to use there.

Most of those libraries (e.g. Gtk, Qt, Panda3D) predate virtualenv/pip - which means you can't just

`pip install panda3d` - since it's setup.py doesn't know about virtualenv.

One way people get round this is with symlinks ... although this is tricky in Windows.

Vext works by modifying the PATH and PYTHONPATH, also it implements a finder/loader, so that only the libraries you allow can break out of the virtualenv.

If you have QT5 installed, you can run:

`pip install vext.qt5` in a virtualenv.    - I haven't tested vext.pyqt5 in Windows, but if it doesn't work there, that's a bug.

[EDIT] And then use QT5 in the virtualenv, just like you can in the system python.  [/EDIT]

**Comment by szabolcsdombi** [comment]
tested with:
- Python33 (32bit)
- Python34 (32bit)
- Python35 (32bit)
- Python36 (32bit)
- Python33 (64bit)
- Python34 (64bit)
- Python35 (64bit)
- Python36 (64bit)
- Python27 (32bit)
- Python27 (64bit)

Checkout the [log](https://gist.github.com/cprogrammer1994/d6f27b8ca22b78be8afae4b73ab03841).

**Comment by stuaxo** [comment]
Wow, cool - how did you test on all the different python versions, using tox or something, or just manually ?

I'm guessing none / not all of those worked from the output.  I'll try testing this in a Windows 10 VM with Qt5 installed.

BTW, because vext links from your virtualenv back to the system python, it will only work with the system python used to create the virtualenv (because the python version must be identical).
- Basically there is a file called 'orig-prefix.txt' in the venv, that has a line that points at the system python.

Vext then looks there for QT5.

If things are working properly you should get similar output from these diagnostic commands:

```
$ vext -l
Running in virtualenv [enabled]
pyqt5.vext
```

```
$ vext -c pyqt5.vext
import sip: [success]
import PyQt5: [success]
```

**Comment by szabolcsdombi** [comment]
I have a ram partition. after every boot i copy and unzip these versions.
i place them in the Path folder inside the ram partition.
there will be the [Pythons.bat](https://gist.github.com/cprogrammer1994/c1a485b329df9dd7f2ebce1aee44a676) file as well.

```
pythons -m pip install ...
```

Personally i use for [uploading](https://github.com/cprogrammer1994/ModernGL/blob/master/multi_upload.bat).

it is a great clean environment for testing. if something broken i just delete and unzip the files again.
every python install has the latest pip installed.

I can help you test your module with these. tell me what to do to get your module ready.

**Comment by stuaxo** [comment]
I'm going to move chat about this to #33 since this bug is about Gtk, but I reckon we can use your thing more generally (and maybe start with PyQt5 as it should work).
